kafkabp/consumer.go

```diff
@@ -0,0 +1,229 @@
+package kafkabp
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// ConsumeMessageFunc is a function type for consuming consumer messages.
+type ConsumeMessageFunc func(ctx context.Context, msg *sarama.ConsumerMessage) error
+
+// ConsumeErrorFunc is a function type for consuming consumer errors.
+type ConsumeErrorFunc func(err error)
+
+// consumer is an instance of a Kafka consumer.
+type consumer struct {
+	cfg ConsumerConfig
+
+	consumer           atomic.Value // sarama.Consumer
+	partitions         atomic.Value // []int32
+	partitionConsumers atomic.Value // []sarama.PartitionConsumer
```

Comment:
@fishy do you know why we are using an atomic value for this?  It seems like it might be because it's read by `Close` in addition to being read/set in `Consume`?

🔕  Is this how we should even be handling this? This seems like it might be an over-use of the `atomic` package which has some pretty strong opinions about when to use it.